Some improvements and fixes, mostly from dhewm3

DanielGibson opened the pull request:
* Some GCC/Clang specific compiler options
    - CMake option to enable ASan
    - CMake option to enforce colored compiler messages
    - Safer options for floating point math
* fixed alignment of memory returned by `idClass::operator new()` (i.e. whenever an entity or other subtype of idClass is allocated with `new`) on 64bit systems by aligning to `intptr_t` (8byte == 64bit on 64bit systems) instead of `int` (4byte == 32bit even on 64bit systems)
* Fix references pointing to invalid stack memory (temporaries that went out of scope)
    - Enable GCC warning about this (if supported by the used GCC version) as error (=> if GCC detects this issue compilation fails). I think GCC 13 and newer support it.

stgatilov commented on CMakeLists.txt:
```diff
@@ -69,6 +71,34 @@ if (UNIX)
+
+	if(FORCE_COLORED_OUTPUT)
+		if(CMAKE_COMPILER_IS_GNUCC)
+		   add_compile_options (-fdiagnostics-color=always)
```
I read about this setting on GCC website, and I don't understand why we should even touch it.
This is 100% user preference, and he can configure his environment how we wants.
Overriding those user settings in the project sound not just unnecessary, but even a bad thing to do, no?

DanielGibson replied:
When building with make GCC and Clang use colors by default, when building with ninja they don't, for some reason, unless you force them to.
This makes it easy to force them without messing with environment variables.

I don't see how this could be a bad thing, it's only done if the user explicitly sets the `FORCE_COLORED_OUTPUT` CMake option, after all.

stgatilov commented on CMakeLists.txt:
```diff
@@ -69,6 +71,34 @@ if (UNIX)
+		add_compile_options(-ffp-contract=off)
+	endif()
+
+	CHECK_CXX_COMPILER_FLAG("-Wdangling-reference" cxx_has_Wdangling_reference)
```
Could you please move it above to line 58?
So that all warning settings were in single place...

DanielGibson replied:
ok

stgatilov commented on game/gamesys/Class.cpp:
```diff
@@ -452,11 +452,11 @@ idClass::new
-	s += sizeof( int );
-	p = (int *)Mem_Alloc(static_cast<int>(s));
-	*p = static_cast<int>(s);
+	s += sizeof( intptr_t );
```
Okay, I guess it makes sense to align pointer members properly =)

stgatilov commented on CMakeLists.txt:
```diff
@@ -69,6 +71,34 @@ if (UNIX)
+	# and https://github.com/RobertBeckebans/RBDOOM-3-BFG/issues/436#issuecomment-850930254 (and following)
+	CHECK_CXX_COMPILER_FLAG("-ffp-contract=off" cxx_has_fp-contract)
+	if(cxx_has_fp-contract)
+		add_compile_options(-ffp-contract=off)
```
Oh, this is an interesting thing.

I wouldn't say that compiler is wrong in optimizing `x*y - y*x` to getting nonzero result, but I agree that it's damn unexpected and inconvenient =) That's exactly what people wanted to avoid when they switched from FPU to SSE2. By the way, the original Doom 3 in this question depended on the mercy of compiler, because when GCC compiles for FPU, there is no way at all to make FPU rounding deterministic...

The current TDM avoids this issue because:
  1. That dmap triangulation optimization function is replaced already with different algorithm.
  2. We never use arch=native and instead only enable SSE2, so compiler just can't use FMA yet.

But I agree that this is a good change anyway.
If I faced FMA contractions myself, I would definitely do the same.

stgatilov commented:
You can commit the changes to SVN.
Merging GitHub PR won't work.

DanielGibson commented:
Ok, I was hoping you'd have something set up to push from git to svn ;)

Once `git svn clone` is done here (seems to take forever, thanks also to boost) I'll try my luck using that for committing. If that fails I'll create a svn commit from the diff, like an animal in the woods...

stgatilov commented:
No, there is nothing automatic yet, not that we have many PRs on GitHub.
I can just copy over the files into SVN working copy and commit =)

If it is possible to clone not from the first revision, you can save yourself on downloading a lot of garbage =)
In the initial times SVN repo was quite a trashbin.

DanielGibson commented:
> If it is possible to clone not from the first revision, you can save yourself on downloading a lot of garbage =)

Yes, one should definitely do this, even if the git-svn documentation isn't very good at mentioning how to do it..

I got it working, this is what I did:

* `git svn clone --username=covert_caedes -r10948:HEAD https://svn.thedarkmod.com/svn/darkmod_src/trunk/ darkmod_gitsvn`
  - Only gets the commits from 10948 to the latest one, that's the "Recorded back merge from 2.13 release branch (no real changes)" commit, the one before "Bumped TDM version to 2.14." - seemed like a good start
  - This took about 24 seconds and afterwards I had a git repo whichs `main` branch corresponds to the darkmod_src SVN trunk (with limited history)
* (in the freshly created `darkmod_gitsvn/` directory) I ran:  
  `git remote add dg git@github.com:DanielGibson/darkmod_src.git`  
  to add my fork of this Github repo, naming it "dg". Then  
  `git fetch dg`  
  to fetch its contents (does not actually check it out, just downloads the (meta)data to `.git/`)
* Then I printed the last 10 commits of my `dg-misc-fixes-rb` **git** branch (the one used by this PR) - 10 because I'm adding less than 10 commits here so that'll list them all and a few existing ones):  
  `git log --oneline dg/dg-misc-fixes-rb -n 10`
  That looks like this:  
  ```
  72c14ff1c (dg/dg-misc-fixes-rb) CMakeLists.txt: Disable -ffast-math, use some alternatives
  eab2dd0b5 CMakeLists.txt: Disable -ffp-contract=fast optimization for GCC/Clang
  98ce016ba idClass::operator new() and delete: align to intptr_t
  6f23b81cb Fix some references pointing to invalid memory
  0f6683237 Fix some references pointing to invalid memory
  d078b6e3f .gitignore: ignore all build*/ directories (not just build/)
  2f9239129 CMake: Add support for ASan and FORCE_COLORED_OUTPUT on Linux
  1741cfd3b (upstream/trunk, upstream/HEAD, dg/trunk, dg/HEAD) #6588: Added fix to prevent infinite decals from spawning when a moveable gets stuck inside breakable glass
  f28ec1b94 #6546,#6623. Minor fix for debug visualization.
  68890a854 #6546,#6623. Properly supported blood markers in Light Estimate System.
  ```
* Just to be sure I also checked the last two commits in the `main` branch (from the SVN `trunk`):  
  `git log --oneline -n2`  
  printing:
  ```
  0e5154caf (HEAD -> main, git-svn) #6588: Added fix to prevent infinite decals from spawning when a moveable gets stuck inside breakable glass
  95b04978e #6546,#6623. Minor fix for debug visualization.
  d67b2cf0f #6546,#6623. Properly supported blood markers in Light Estimate System.
  ```
* Ok, looks fine (the commit IDs are obviously different between my local `main` from my git-svn clone and my git branch based on this Github repo here - that's expected, but the commit messages line up).  
  So I want to add all the commits from `2f9239129 CMake: Add support for ASan and FORCE_COLORED_OUTPUT on Linux` on until `72c14ff1c  CMakeLists.txt: Disable -ffast-math, use some alternatives`
* Now **cherry-pick** these commits from my Git branch to the `main` (git-svn) branch:  
  `2f9239129^..72c14ff1c`  
  - **Note** the `^` after the first commit, in gits weird commit-range syntax that is required to actually include that first commit: `A..B` means "the commits **after** A up to including B", `A^..B` means "the commits from A up to including B" (or actually "the commits **after** A's parent commit up to including B").
* Run `git log --oneline` to confirm that all the commits were added to the main branch.
* Commit this to svn:  
  `git svn dcommit`
* Then I updated my regular SVN checkout of the repo (with `svn up`), checked its log (`svn log | less`) and realized that it worked as expected..  
  *.. but I'm a fucking idiot and missed the duplicate "Fix some references pointing to invalid memory" commit in all the previous steps (it moved something around in CMakeLists.txt as requested in this review, the commits were supposed to be squashed into one but I somehow screwed that up)*

This seems to work well overall :)
One step that I must do in the future before cherry-picking new commits is running `git svn rebase`, which gets the latest changes from the svn repo (basically does `svn up`). If any local commits exist that haven't been pushed to svn yet, they will be added after the latest commit from the repo (like `git pull --rebase`).

DanielGibson commented:
BTW: Thanks for the review! :)
